connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java

```diff
@@ -344,12 +373,32 @@ private void rewind() {
         context.clearOffsets();
     }
 
+    private void openPartitions(Collection<TopicPartition> partitions) {
+        if (partitions.isEmpty())
+            return;
+
+        task.open(partitions);
+    }
+
+    private void closePartitions() {
+        if (currentOffsets.isEmpty())
+            return;
+
+        Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>(currentOffsets);
+        try {
+            task.flush(offsets);
+        } finally {
+            // Some sink implementations may not actually flush the data until close is called,
+            // so the commit should follow it.
+            task.close(currentOffsets.keySet());
+        }
```

Comment:
There's definitely some inconsistency here. The goal was to ensure that we don't try to flush partitions that have already been closed. However, as you've pointed out, this makes the commit callback handling a little weird. If flush() fails, the callback won't be triggered, but it will be if commit fails. We should probably make this handling consistent by catching the exception raised from flush() and invoking onCommitCompleted(). 

As for the sequence number, I'm fairly sure that the consumer preserves the order of commits, so it may not be necessary. If the commit is synchronous, the consumer waits for its completion. If it's asynchronous, we give up on the first error. You're guaranteed after a call to commitAsync that the request has either been transmitted or has failed, and the broker guarantees that requests are handled in order. 

Because of the current use of "offset.flush.timeout.ms," it is possible that multiple commits get queued up on the broker. Since async commits are guaranteed to complete by the expiration of the consumer's request timeout, there could be as many as `request.timeout.ms/offset.flush.timeout.ms` outstanding commits. The sequence number deals with this basically by letting the task "ignore" commits which are already considered failed. If I understand correctly, what you're proposing as an alternative is to pause processing until the commit completes. Then we definitely don't need the sequence number since we only allow one commit on the wire. It seems like that could work. If a commit is timing out, there's a good chance of a problem on the coordinator and our next commits will timeout also. So basically it keeps the sink task from getting too far ahead of the last successful commit, which seems like a nice property.

Maybe we can leave that for another JIRA to keep this patch from getting any bigger? For now, is there any reason why we don't just use the next sequence number in the synchronous commit call? If an async commit is still on the wire, it will return first and we'll ignore it, which seems ok. Then when the sync commit comes back, we'll hit the right path.